### ISSUE TITLE
Feedback for VERSION_CATALOGS: dynamically added dependencies not configurable

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#18979]
When using `enableFeaturePreview("VERSION_CATALOGS")` imagine I'm adding a dependency for `g[roup]:a[rtifact]:v[ersion]`.

Without catalogs, this works:
```kotlin
add("compileOnly", "g:a:v") { isTransitive = false }
```
with a catalog:
```toml
[libraries]
g-a = { module = "g:a", version = "v" }
```
and build.gradle.kts:
```kotlin
add("compileOnly", libs.g.a) { isTransitive = false }
```
doesn't.

### Expected Behavior
compiles and works, same as `add("compileOnly", libs.g.a)` does.

### Current Behavior
```
> Configure project :
e: build.gradle.kts:136:40: Type mismatch: inferred type is () -> Unit but Closure<Any> was expected
e: build.gradle.kts:137:5: Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:
public val <T : Configuration> NamedDomainObjectProvider<TypeVariable(T)>.isTransitive: Boolean defined in org.gradle.kotlin.dsl

FAILURE: Build failed with an exception.

* Where:
Build file 'build.gradle.kts' line: 136

* What went wrong:
Script compilation errors:

  Line 136:                     add("compileOnly", libs.g.a) {
                                                   ^ Type mismatch: inferred type is () -> Unit but Closure<Any> was expected

  Line 137:                             isTransitive = false
                ^ Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:
                    public val <T : Configuration> NamedDomainObjectProvider<TypeVariable(T)>.isTransitive: Boolean defined in org.gradle.kotlin.dsl

2 errors
```

#### addProvider
Which is a mouthful and inconsistent with the working `add("compileOnly", libs.g.a)`.
```
@Suppress("UnstableApiUsage")
addProvider<MinimalExternalModuleDependency, ExternalModuleDependency>("compileOnly", libs.g.a) {
	isTransitive = false
}
```

#### buildSrc polyfill
```kotlin
/**
 * Polyfill for not yet added method.
 * TODEL https://github.com/gradle/gradle/issues/18979
 */
fun DependencyHandlerScope.add(
	configurationName: String,
	dependency: Provider<MinimalExternalModuleDependency>,
	configuration: Action<in ExternalModuleDependency>
) {
	addProvider(configurationName, dependency, configuration)
}
```

### Context
I'm simply trying to migrate my existing dependencies to toml. In some cases it's necessary to use `add(conf, dep)` notation, rather than `conf(dep)`. Gradle catalogs break down in this case.

### Steps to Reproduce 
1.
2. run `gradlew`

### Your Environment

Build scan URL: N/A

**Comment by ljacomet** [comment]
@octylFractal Is this something you took care of already?

**Comment by vlsi** [comment]
I believe the issue duplicates https://github.com/gradle/gradle/issues/18617

**Comment by octylFractal** [comment]
@ljacomet No, also this is not strictly version-catalog related, it's a general deficiency in the Kotlin DSL that needs improvement.

Workaround for now is to use the alternative invoke paths, e.g. `"compileOnly"(libs.g.a) { isTransitive = false }`. This will not work for ProviderConvertible until 7.4 (that's the duplicate @vlsi linked). However, that does not cover fixing specifically `add`, and there needs to be a discussion about the best path forward to create a smoother experience for these edge-case APIs.

**Comment by TWiStErRob** [comment]
Yeah, sounds like this might be in the scope of the fix of #18617. I have a question why the different add, addProvider, addPC methods? Why not just add overloaded with different params and let the compiler choose?

**Comment by octylFractal** [comment]
IIRC they were added by @melix, I know there was some reasoning about Java plugin use-cases, but there is currently nothing that couldn't be served by the normal `add` APIs if an `add(String, Object, Action)` was added. This is part of what I was referring to with "the best path forward to create a smoother experience" -- these APIs aren't consistently designed and need some refinement.

**Comment by melix** [comment]
Basically the underlying reason is that the `add` API is very old and uses `Closure` based actions which are not "strongly typed" enough to give static languages (Java, Kotlin) enough context for completion. That is, in a build script you want to be able to infer the type of the lambda delegate. This is what `addProvider` offers, which wasn't doable using `add`. There are some technical difficulties in migrating `add` to `Action` which I don't precisely remember (it's a few months old already).

**Comment by octylFractal** [comment]
Unfortunately `addProvider` currently is overly specific, as it assumes everything returns an `ExternalModuleDependency`, which is not true in many cases. Typing that better is difficult in Java though, as erasure prevents us from differentiating between e.g. `Provider<String>` and `Provider<T extends Dependency>`. I do want to dive deeper into this in the future to figure out a better solution.

**Comment by melix** [comment]
That's right. It was an intentional decision precisely because of what you mention (no way to differentiate because of type erasure). Unfortunately I'm not sure we can do much better without having to add specific methods with specific names for each case.

**Comment by TWiStErRob** [comment]
Makes sense, erasure 🤦🏻‍♂️, but then something is still fishy about the workaround I mentioned, why do I need to specify both generics explicitly? Anyway, I like https://github.com/gradle/gradle/issues/18979#issuecomment-969253055, let's hope that lands soon!

**Comment by octylFractal** [comment]
The reason is that without it, the `U` type parameter cannot be inferred. You should be able to type the lambda instead:
```kotlin
addProvider("compileOnly", libs.g.a, Action<ExternalModuleDependency> {
    isTransitive = false
})
```
Unfortunately it doesn't seem that Kotlin has syntax to specify the type of the receiver without using the more verbose SAM conversion syntax.

**Comment by vlsi** [comment]
```kotlin
addProvider("compileOnly", libs.g.a, Action<ExternalModuleDependency> {
    isTransitive = false
})
```

> prevents us from differentiating between e.g. `Provider<String>` and `Provider<T extends Dependency>`.

Can you please clarify why do you want to differentiate between `Provider<String>` and `Provider<T extends Dependency>`?

> The reason is that without it, the U type parameter cannot be inferred

Can you please clarify why is `U` parameter needed in the first place?

Is the current

```java
<T, U extends ExternalModuleDependency> void addProvider(
    String configurationName, Provider<T> dependencyNotation,
    Action<? super U> configuration);
```

different from

```java
void addProvider(
    String configurationName, Provider<?> dependencyNotation,
    Action<? super ExternalModuleDependency> configuration);
```

?

**Comment by octylFractal** [comment]
> Can you please clarify why do you want to differentiate between `Provider<String>` and `Provider<T extends Dependency>`?

A `Provider<String>`, in a perfect world, would type the `Action` as `Action<? super ExternalModuleDependency>`, and a `Provider<T extends Dependency>` would type the `Action` as `Action<? super T>`. This is important for dependencies like `project()` and `files()`, which do not result in `ExternalModuleDependency` objects. If it would be too much to give individual methods for each dependency kind, then we could only offer `Action<? super Dependency>`, as that is the only known common type. Having `addProvider` (effectively) take `Provider<?>`  and return `Action<? super ExternalModuleDependency>` prevents any `Provider<?>` passed from returning `project()`, `files()`, etc. at runtime, and there isn't really a good workaround right now.

> Can you please clarify why is `U` parameter needed in the first place?

It seems unnecessary to me, but I didn't write this API.
